### ISSUE TITLE
lucene delay with first search

### ISSUE THREAD

**Issue by NCIEVS** [NCIEVS/nci-protege5#547]
Observed in 3.8.0 qa2.  From [NT]

"First search of the day can take minutes to return results. All searches after that or normal."

Advised NT:

"info that would be good to have are a) a CPU check with the taskmanager when this happens to see if the hiccup is from some other windows program . . . , b) a thread dump (control-break in protege's terminal window) when this happens."

**Comment by NCIEVS** [comment]
Additional info from [NT]

It happened also on 3.9.0, yesterday when the ticket was created.

And the indexing progress bar does not show up when this happens.

**Comment by NCIEVS** [comment]
From [NT, 02/18/2022 16:34:33]

Search hung at 4:31 PM.  

Log is consistent with the previous logs, index updating with changesets at the time of the hiccup.

**Comment by bdionne** [comment]
This is working as intended. When a project is opened the last thing that's executed is the update of the search indices. This is done in a separate thread, so for users who wish to proceed and edit there's no need to wait.

If the user immediately execute a search it won't be executed until the update is complete. We could change this behavior a couple of ways:

 * Not use a separate thread for the index update, so that the user would be forced to wait for it to complete before the project is completely open
 * Use an additional separate thread for search, allowing the search to execute in parallel. This would require some code changing but presumably could be made to work. However the search results would be incomplete as they would not incorporate the latest changes.
 
For now it seems the very simple workaround, knowing that this is what's happening, is to wait a few minutes before searching.

**Comment by bdionne** [comment]
This can be fixed provided users can deal with possibly incomplete results on first search. The indexing of change sets on open project and be done in a background thread, allowing a search thread to proceed.

**Comment by singhik** [comment]
@bdionne : What is the 'Fix' applied for this ticket ? Is Search a separate thread now ? is Searching not part of INdexing now ? , How does Search and INdexing work now after this fix ? 
How does concurrency of user(s) impacted ? A real time use case ?

**Comment by bdionne** [comment]
@singhik

> lucene delay with first search

Presumably you tested this delay and reproduced it? Have you tried it now, is there still a delay? 

> What is the 'Fix' applied for this ticket?

The fix is to not allow searching until the indexing is complete. [this](https://github.com/NCIEVS/lucene-search-tab/commit/58f753254bf612f154675d04c41a9d431cf5c3d7) is the detail, happy to discuss. We discussed this last week in status, I can explain it again if you like.

> Is Search a separate thread now ? 
No

> How does concurrency of user(s) impacted ?
Not sure what you mean, but you should *test* anything you think might be impacted

> How does Search and INdexing work now after this fix ?
Much better, and no change to how one uses it, it's all the same, just no latency

Again, as I mentioned in status: The user reported a long delay in the first search after bringing up the application. This was due to the change sets from the server being indexed. That delay no longer occurs. When the open dialog goes away you can begin searching. Give it a try.

Let me know if this answers your questions

**Comment by singhik** [comment]
@bdionne @NCIEVS  : Can't perform a real-time search for added concepts . Steps : 

Add/Create a Concept : 

![image](https://user-images.githubusercontent.com/36314371/163605602-0a019896-0d11-4c89-92b5-d4fcbf11a073.png)

Perform Search using label : Unable to Search without manual indexing

![image](https://user-images.githubusercontent.com/36314371/163605650-a9fb4ff9-2954-4aae-8a64-a3750ecd8fbe.png)

![image](https://user-images.githubusercontent.com/36314371/163605670-008c60af-67b5-4ca5-aa12-bda7054c81d9.png)


Auto-Update is checked : 

![image](https://user-images.githubusercontent.com/36314371/163605717-5bfe84de-c341-4c9f-9615-9147ef8507ce.png)

**Comment by bdionne** [comment]
This is probably the same issue as #552 also, and I will look at this when I'm finished testing the other two

@singhik The issue to be testing here is the delay on the first search *after* bringing up the app.

**Comment by singhik** [comment]
@bdionne  : The scenario when Project loads and Search taking long has been validated , also validated a scenario to try Search when Index is running , both these scenarios work fine and do not allow any Search unless Indexing is completed . 

The one above ,where it failed , is also part of this test

**Comment by bdionne** [comment]
@singhik good to know something works :) This new issue is the same as the other two tickets, so hopefully all three are fixed together

**Comment by bdionne** [comment]
This is fixed in [nci-edit-tab-1.4.0](https://github.com/NCIEVS/nci-edit-tab/releases/tag/nci-edit-tab-1.4.0). It requires [lucene-search-tab-1.3.0](https://github.com/NCIEVS/lucene-search-tab/releases/tag/lucene-search-tab-1.3.0) and the protege [4.0.0-dev2](https://github.com/NCIEVS/protege/releases/tag/4.0.0-dev2) build.

This also fixes #552 and #434

**Comment by singhik** [comment]
@NCIEVS @bdionne  : Fails on nci-edit-tab 1.4.0 , lucene-search 1.3.0 and 4.0.0-dev2 build  

Create a Concept 'Test4' : 

![image](https://user-images.githubusercontent.com/36314371/164993605-b127663f-5f69-4c29-950e-f3f16e31956d.png)

Login to another instance of same client,same Project and different user and perform Search of above concept(Keeps searching in infinite loop and does not stop , clicking on 'Stop Search' does not work either : ) : 

![image](https://user-images.githubusercontent.com/36314371/164993642-054ec7e3-2acd-4318-93ae-a241967ad392.png)

![image](https://user-images.githubusercontent.com/36314371/164993734-e7fdc44d-60af-49a3-b5c2-9db1bd46e712.png)


Hierarchy Tree of second instance does show 'Test4' indicating the project has updated data : 

![image](https://user-images.githubusercontent.com/36314371/164993719-7c1cc3f1-e78b-47bf-b1bd-cdb7399021f3.png)

![image](https://user-images.githubusercontent.com/36314371/164993747-0898b109-aa0e-41f7-bf24-e0b29839fb7a.png)

**Comment by bdionne** [comment]
> Login to another instance of same client,same Project and different user and perform Search of above concept(Keeps searching in infinite loop and does not stop , clicking on 'Stop Search' does not work either

@singhik - Is this second instance on the same machine? If so and it's the same install of the client then it will share the search indices and I believe it will lock the files. Please clarify if this is the case, and if not I'll look into this.

cc: @NCIEVS

**Comment by singhik** [comment]
@bdionne : Same Machine and different folder locations

**Comment by bdionne** [comment]
You must have a pretty large RAM size to run 2 instances of protege loading Thesaurus on the same machine.

The first time you open each instance it should take some time to index, but subsequent runs should be faster. You should also see any edits in one instance show up in the second, once you click the updates button. 

I'll look into it. Thanks.

**Comment by singhik** [comment]
The scenario of concurrent users in a same machine is not a real-time simulation hence there are dependencies on second instance to execute indexing for changes to be Lucene searched from first instance. Concurrency works well in @NCIEVS 's system. Hence this ticket is closed.
